server/src/main/java/org/elasticsearch/gateway/GatewayMetaState.java

```diff
@@ -73,44 +79,64 @@
 
     //there is a single thread executing updateClusterState calls, hence no volatile modifier
     protected Manifest previousManifest;
     protected ClusterState previousClusterState;
+    protected boolean clusterStateUpdatersApplied;
     protected boolean incrementalWrite;
 
     public GatewayMetaState(Settings settings, NodeEnvironment nodeEnv, MetaStateService metaStateService,
-                            MetaDataIndexUpgradeService metaDataIndexUpgradeService, MetaDataUpgrader metaDataUpgrader) throws IOException {
+                            MetaDataIndexUpgradeService metaDataIndexUpgradeService, MetaDataUpgrader metaDataUpgrader,
+                            TransportService transportService, ClusterService clusterService,
+                            IndicesService indicesService) throws IOException {
         this.settings = settings;
         this.nodeEnv = nodeEnv;
         this.metaStateService = metaStateService;
+        this.transportService = transportService;
+        this.clusterService = clusterService;
+        this.indicesService = indicesService;
 
         ensureNoPre019State(); //TODO remove this check, it's Elasticsearch version 7 already
         ensureAtomicMoveSupported(); //TODO move this check to NodeEnvironment, because it's related to all types of metadata
         upgradeMetaData(metaDataIndexUpgradeService, metaDataUpgrader);
         initializeClusterState(ClusterName.CLUSTER_NAME_SETTING.get(settings));
         incrementalWrite = false;
+        clusterStateUpdatersApplied = false;
     }
 
     private void initializeClusterState(ClusterName clusterName) throws IOException {
         long startNS = System.nanoTime();
         Tuple<Manifest, MetaData> manifestAndMetaData = metaStateService.loadFullState();
         previousManifest = manifestAndMetaData.v1();
 
+        final MetaData metaData = manifestAndMetaData.v2();
+        final ClusterBlocks.Builder blocks = ClusterBlocks.builder().addGlobalBlock(STATE_NOT_RECOVERED_BLOCK);
+
         previousClusterState = ClusterState.builder(clusterName)
                 .version(previousManifest.getClusterStateVersion())
-                .metaData(manifestAndMetaData.v2()).build();
+                .blocks(blocks.build())
+                .metaData(metaData).build();
 
         logger.debug("took {} to load state", TimeValue.timeValueMillis(TimeValue.nsecToMSec(System.nanoTime() - startNS)));
     }
 
-    public void setLocalNode(DiscoveryNode localNode) {
-        assert previousClusterState.nodes().getLocalNode() == null : "setLocalNode must only be called once";
-        previousClusterState = ClusterState.builder(previousClusterState)
-                .nodes(DiscoveryNodes.builder().add(localNode).localNodeId(localNode.getId()).build())
-                .build();
+    public void applyClusterStateUpdaters() {
+        assert clusterStateUpdatersApplied == false : "applyClusterStateUpdaters must only be called once";
```

Comment:
I prefer the `previousClusterState.nodes().getLocalNode() == null` to adding another mutable field to this class.
For extra safety, you can assert that `transportService.getLocalNode() != null` when this is called.